src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/MemorySegment.java

```diff
@@ -693,165 +581,175 @@ default MemorySegment asSlice(MemoryAddress newBase) {
      *
      * @param arr the primitive array backing the array memory segment.
-     * @return a new confined array memory segment.
+     * @return a new array memory segment.
      */
     static MemorySegment ofArray(long[] arr) {
         return HeapMemorySegmentImpl.OfLong.fromArray(arr);
     }
 
     /**
      * Creates a new confined array memory segment that models the memory associated with a given heap-allocated double array.
-     * <p>
-     * The resulting memory segment keeps a reference to the backing array, to ensure it remains <em>reachable</em>
-     * for the life-time of the segment. The segment will feature all <a href="#access-modes">access modes</a>
-     * (see {@link #ALL_ACCESS}).
+     * The returned segment's resource scope is set to the {@link ResourceScope#globalScope() global} resource scope.
      *
      * @param arr the primitive array backing the array memory segment.
-     * @return a new confined array memory segment.
+     * @return a new array memory segment.
      */
     static MemorySegment ofArray(double[] arr) {
         return HeapMemorySegmentImpl.OfDouble.fromArray(arr);
     }
 
     /**
-     * Creates a new confined native memory segment that models a newly allocated block of off-heap memory with given layout.
+     * Creates a new confined native memory segment that models a newly allocated block of off-heap memory with given layout
+     * and resource scope. A client is responsible make sure that the resource scope associated to the returned segment is closed
+     * when the segment is no longer in use. Failure to do so will result in off-heap memory leaks.
      * <p>
      * This is equivalent to the following code:
      * <blockquote><pre>{@code
-    allocateNative(layout.bytesSize(), layout.bytesAlignment());
+    allocateNative(layout.bytesSize(), layout.bytesAlignment(), scope);
      * }</pre></blockquote>
-     *
-     * @implNote The block of off-heap memory associated with the returned native memory segment is initialized to zero.
-     * Moreover, a client is responsible to call the {@link MemorySegment#close()} on a native memory segment,
-     * to make sure the backing off-heap memory block is deallocated accordingly. Failure to do so will result in off-heap memory leaks.
+     * <p>
+     * The block of off-heap memory associated with the returned native memory segment is initialized to zero.
      *
      * @param layout the layout of the off-heap memory block backing the native memory segment.
+     * @param scope the segment scope.
      * @return a new native memory segment.
      * @throws IllegalArgumentException if the specified layout has illegal size or alignment constraint.
      */
-    static MemorySegment allocateNative(MemoryLayout layout) {
+    static MemorySegment allocateNative(MemoryLayout layout, ResourceScope scope) {
+        Objects.requireNonNull(scope);
```

Comment:
Should the allocateNative methods declare that they throw ISE, if the given ResourceScope is not alive?   ( I found myself asking this q, then considering the behaviour of a SegmentAllocator that is asked to allocate after a RS has been closed )

Reply:
Good point, yes it should